keps/sig-network/20200130-Uncapped-IPv6.md

```diff
@@ -0,0 +1,316 @@
+---
+title: Uncapped IPv6
+authors:
+  - "@aojea"
+owning-sig: sig-network
+participating-sigs:
+  - sig-scalability
+  - sig-api-machinery
+reviewers:
+  - "@thockin"
+  - "@bowei"
+  - "@liggitt"
+  - "@smarterclayton"
+  - "@wojtek-t"
+approvers:
+  - "@thockin"
+  - "@liggitt"
+  - "@wojtek-t"
+editor: TBD
+creation-date: 2020-01-30
+last-updated: 2020-01-30
+status: provisional
+see-also:
+  - "/keps/sig-network/20190714-IPv6-beta-proposal.md"
+  - "https://github.com/kubernetes/kubernetes/issues/44918"
+  - "https://github.com/kubernetes/kubernetes/pull/86620"
+  - "https://github.com/kubernetes/kubernetes/pull/79993"
+
+---
+
+# Remove IPv6 Allocation Restrictions
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Roaring bitmaps](#roaring-bitmaps)
+  - [Services IP Allocator](#services-ip-allocator)
+  - [Node IPAM Controller and Range Allocator](#node-ipam-controller-and-range-allocator)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes impose some restrictions on the number of Service IPs and CIDR Range sizes when using IPv6.
+This is done to avoid scalability issues caused by IPv6, we should take into account that IPv4 has an
+address namespace of 2^32 addresses vs 2^128 with IPv6.
+
+## Motivation
+
+One of the main motivation to restrict allocations when using IPv6 is caused because Kubernetes is using internally bitmaps (or bitsets) to store the data.
+Bitmaps are fast data structures, but they can use too much memory and not efficiently. Per example, in sparse scenarios like with the ip allocator,
+if we allocate one IP at the position 1,000,000 we have to use over 100kB only for that IP.
+
+The IPv6 allocation restrictions can be removed or relaxed if we use compressed bitmaps, it also will reduce the size of the data needed to store in etcd.
```

Comment:
Is the restriction actually problematic? Do people really want to allocate so many service IPs?

Reply:
The only problem that I see right now is that it imposes a limitation in your subnet planning, not a big deal though, but don't know what are the limits that users are having with the limit service IPs either and if this can alleviate them. 
I think that at least is a good exercise sharing and discussing, but I personally totally agree with you here https://github.com/kubernetes/enhancements/pull/1534/files#r378460244

Reply:
I hear more cases of "I need more nodeports" or "I want LESS service IPs".  I don't think I have every had a case of needing more than /16 or /112 for services